common/persistence/pinot/pinot_visibility_metric_clients_test.go

```diff
@@ -0,0 +1,1220 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package pinotvisibility
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+
+	"github.com/uber/cadence/.gen/go/indexer"
+	"github.com/uber/cadence/common/definition"
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/log/testlogger"
+	"github.com/uber/cadence/common/metrics"
+	metricsClientMocks "github.com/uber/cadence/common/metrics/mocks"
+	"github.com/uber/cadence/common/mocks"
+	p "github.com/uber/cadence/common/persistence"
+	pnt "github.com/uber/cadence/common/pinot"
+	"github.com/uber/cadence/common/service"
+	"github.com/uber/cadence/common/types"
+)
+
+var (
+	testStopwatch = metrics.NoopScope(metrics.PinotRecordWorkflowExecutionStartedScope).StartTimer(metrics.PinotLatency)
+)
+
+func TestMetricClientRecordWorkflowExecutionStarted(t *testing.T) {
+	// test non-empty request fields match
+	errorRequest := &p.RecordWorkflowExecutionStartedRequest{
+		WorkflowTypeName: "errorWorkflowTypeName",
+		Memo: &types.Memo{
+			map[string][]byte{
+				"CustomStringField": []byte("test string"),
+			},
+		},
+	}
+
+	request := &p.RecordWorkflowExecutionStartedRequest{
+		WorkflowTypeName: "wtn",
+		Memo: &types.Memo{
+			map[string][]byte{
+				"CustomStringField": []byte("test string"),
+			},
+		},
+	}
+
+	tests := map[string]struct {
+		request       *p.RecordWorkflowExecutionStartedRequest
+		expectedError error
+	}{
+		"Case1: error case": {
+			request:       errorRequest,
+			expectedError: &types.BadRequestError{},
+		},
+		"Case2: normal case": {
+			request:       request,
+			expectedError: nil,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			// create mock clients
+			ctrl := gomock.NewController(t)
+			mockPinotClient := pnt.NewMockGenericClient(ctrl)
+			mockProducer := &mocks.KafkaProducer{}
+			mockMetricClient := &metricsClientMocks.Client{}
+			mockScope := &metricsClientMocks.Scope{}
+
+			// create metricClient
+			logger := testlogger.New(t)
+			mgr := NewPinotVisibilityStore(mockPinotClient, &service.Config{
+				ValidSearchAttributes:  dynamicconfig.GetMapPropertyFn(definition.GetDefaultIndexedKeys()),
+				ESIndexMaxResultWindow: dynamicconfig.GetIntPropertyFn(3),
+			}, mockProducer, testlogger.New(t))
+			visibilityStore := mgr.(*pinotVisibilityStore)
+			pinotVisibilityManager := p.NewVisibilityManagerImpl(visibilityStore, logger)
+			visibilityMgr := NewPinotVisibilityMetricsClient(pinotVisibilityManager, mockMetricClient, logger)
+			metricsClient := visibilityMgr.(*pinotVisibilityMetricsClient)
+
+			// mock behaviors
+			mockMetricClient.On("Scope", mock.Anything, mock.Anything).Return(mockScope).Once()
+			mockScope.On("StartTimer", mock.Anything, mock.Anything).Return(testStopwatch).Once()
```

Comment:
This initialisation part looks pretty much the same across all 17 tests which is a big overhead.
Is it possible to extract it into some function?
This looks like a great candidate to Suite. I know we're trying to avoid it, but we can maybe borrow the idea of TestSuite' `SetupTest()`.

Reply:
Moving the absolutely-fixed parts to helper functions make sense but I'd be cautious of doing that because you may end up passing bunch of flags/parameters to make such setup helper functions handle all cases. Some level of redundancy in test code is OK and readability/self-containedness/explicitness/repeatability should be optimized IMO.

Suites work fine until they don't :) When you want to add a new test cases that initializes common components slightly differently you end up doing hacky things ([example](https://github.com/uber-go/cadence-client/pull/1287/files#r1388175996)) or add the redundant code at that point anyway.

Reply:
I would second Taylan's point that it's hard to generalize assertions in a single place, but I think you're right that this can and perhaps should be less verbose, and, to your wider point, that we need to be careful to ensure the tests remain high-quality as we sprint for coverage. 

I didn't see it too concerning, but in general I'd err towards avoiding `if/else` blocks with branching mock assertions such as line 109/116 and subsequently. If the assertions are different between table test invocations, then probably best to set them per table-row (with a func which is invoked to setup the mocks per row). That would make it easier to follow I think.